ansible_base/authentication/utils/claims.py

```diff
@@ -352,58 +401,279 @@ def create_organizations_and_teams(results) -> None:
+
+                for content_type_id, content_type_permissions in role_permissions.items():
+                    for _object_id, object_with_status in content_type_permissions.items():
+                        self._apply_permission(object_with_status, role_name)
+
+    def _apply_permission(self, object_with_status, role_name):
+        status = object_with_status['status']
+        obj = object_with_status['object']
+
+        if status == self.permissions_cache.STATUS_ADD:
+            self._give_permission(self.permissions_cache.rd_by_name(role_name), obj)
+        elif status == self.permissions_cache.STATUS_REMOVE:
+            self._remove_permission(self.permissions_cache.rd_by_name(role_name), obj)
+        elif status == self.permissions_cache.STATUS_EXISTING and self.rebuild_user_permissions:
+            self._remove_permission(self.permissions_cache.rd_by_name(role_name), obj)
+
+    @staticmethod
+    def _get_orgs_by_name(org_names) -> dict[str, AbstractOrganization]:
+        orgs_by_name = {org.name: org for org in Organization.objects.filter(name__in=org_names)}
+        return orgs_by_name
+
+    @staticmethod
+    def _get_teams_by_name(org_id, team_names) -> dict[str, AbstractTeam]:
+        # FIXME(cutwater): Load all teams in all organizations at once.
+        #       This will require raw query to filter by tuples of (org id, team name).
+        teams_by_name = {team.name: team for team in Team.objects.filter(organization__pk=org_id, name__in=team_names)}
+        return teams_by_name
+
+    def _give_permission(self, role_definition: CommonModel, obj: Union[AbstractOrganization, AbstractTeam, None] = None) -> None:
+        if obj:
+            logger.info(
+                _("Assigning role '{rd}' to user '{username}' in '{object}").format(
+                    rd=role_definition.name, username=self.user.username, object=obj.__class__.__name__
+                )
+            )
+        else:
+            logger.info(_("Assigning role '{rd}' to user '{username}'").format(rd=role_definition.name, username=self.user.username))
+
+        if obj:
+            role_definition.give_permission(self.user, obj)
+        else:
+            role_definition.give_global_permission(self.user)
+
+    def _remove_permission(self, role_definition: CommonModel, obj: Union[AbstractOrganization, AbstractTeam, None] = None) -> None:
+        if obj:
+            logger.info(
+                _("Removing role '{rd}' from user '{username}' in '{object}").format(
+                    rd=role_definition.name, username=self.user.username, object=obj.__class__.__name__
+                )
+            )
+        else:
+            logger.info(_("Removing role '{rd}' to user '{username}'").format(rd=role_definition.name, username=self.user.username))
+
+        if obj:
+            role_definition.remove_permission(self.user, obj)
+        else:
+            role_definition.remove_global_permission(self.user)
+
+
+class RoleUserAssignmentsCache:
```

Comment:
This is making me think should we have an RBAC cache inside of Redis for all nodes to use. We would need a good cache-busting strategy to do that though. Not something specifically for this PR.

Reply:
it would need quite more effort indeed